File: testsprite-tests/be-operations.py
import requests
import time
import random

TARGET_URL = "https://yemen-telecom.onrender.com"

def login():
    csrf = requests.get(f"{TARGET_URL}/api/csrf-token", timeout=30).json()
    r = requests.post(f"{TARGET_URL}/api/auth/login", json={"username": "manager", "password": "Admin@123"},
        headers={"X-CSRF-Token": csrf["token"], "X-CSRF-Hash": csrf["hash"]}, timeout=30)
    return r.json()["token"]

def csrf_headers(token):
    csrf = requests.get(f"{TARGET_URL}/api/csrf-token", timeout=30).json()
    return {"Authorization": f"Bearer {token}", "X-CSRF-Token": csrf["token"], "X-CSRF-Hash": csrf["hash"]}

def auth_only(token):
    return {"Authorization": f"Bearer {token}"}

def test_list_operations():
    token = login()
    r = requests.get(f"{TARGET_URL}/api/operations", headers=auth_only(token), timeout=30)
    assert r.status_code == 200, f"Expected 200, got {r.status_code}"

def test_create_operation_activate():
    token = login()
    target = f"77{random.randint(10000000,99999999)}"
    r = requests.post(f"{TARGET_URL}/api/operations", json={"type": "activate", "target": target, "operator": "Yemen Mobile"}, headers=csrf_headers(token), timeout=30)
    assert r.status_code == 201, f"Expected 201, got {r.status_code}: {r.text}"
    assert r.json()["type"] == "activate"

def test_create_operation_missing_target():
    token = login()
    r = requests.post(f"{TARGET_URL}/api/operations", json={"type": "activate"}, headers=csrf_headers(token), timeout=30)
    assert r.status_code == 400, f"Expected 400, got {r.status_code}"

def test_create_operation_invalid_type():
    token = login()
    r = requests.post(f"{TARGET_URL}/api/operations", json={"type": "invalid", "target": "771234567"}, headers=csrf_headers(token), timeout=30)
    assert r.status_code == 400, f"Expected 400, got {r.status_code}"

def test_list_inventories():
    token = login()
    r = requests.get(f"{TARGET_URL}/api/inventories", headers=auth_only(token), timeout=30)
    assert r.status_code == 200, f"Expected 200, got {r.status_code}"
    assert isinstance(r.json(), list)

def test_update_inventories():
    token = login()
    get = requests.get(f"{TARGET_URL}/api/inventories", headers=auth_only(token), timeout=30)
    inventories = [{"operator": inv["operator"], "available": inv.get("available", 0), "remaining": inv.get("remaining", 0)} for inv in get.json()]
    if inventories:
        r = requests.put(f"{TARGET_URL}/api/inventories", json=inventories, headers=csrf_headers(token), timeout=30)
        assert r.status_code == 200, f"Expected 200, got {r.status_code}"

def test_update_inventories_empty():
    token = login()
    r = requests.put(f"{TARGET_URL}/api/inventories", json=[], headers=csrf_headers(token), timeout=30)
    assert r.status_code == 400, f"Expected 400, got {r.status_code}"

test_list_operations()
test_create_operation_activate()
test_create_operation_missing_target()
test_create_operation_invalid_type()
test_list_inventories()
test_update_inventories()
test_update_inventories_empty()

File: testsprite-tests/be-operations.js
const BASE = 'https://yemen-telecom.onrender.com';

async function request(path, options = {}) {
  const res = await fetch(`${BASE}${path}`, {
    headers: { 'Content-Type': 'application/json', ...options.headers },
    ...options,
  });
  const body = res.headers.get('content-type')?.includes('json') ? await res.json() : await res.text();
  return { status: res.status, body, headers: Object.fromEntries(res.headers) };
}

async function test(name, fn) {
  try {
    await fn();
    console.log(`PASS: ${name}`);
  } catch (e) {
    console.error(`FAIL: ${name} — ${e.message}`);
    process.exitCode = 1;
  }
}

function assert(condition, msg) {
  if (!condition) throw new Error(msg || 'Assertion failed');
}

let managerToken;

async function loginManager() {
  const { body } = await request('/api/auth/login', {
    method: 'POST',
    body: JSON.stringify({ username: 'manager', password: 'Admin@123' }),
  });
  managerToken = body.token;
}

async function run() {
  await loginManager();
  if (!managerToken) {
    console.error('Cannot run Operations tests: manager login failed');
    process.exitCode = 1;
    return;
  }

  const auth = { Authorization: `Bearer ${managerToken}` };

  await test('GET /api/operations returns array', async () => {
    const { status, body } = await request('/api/operations', { headers: auth });
    assert(status === 200, `Expected 200, got ${status}`);
    assert(Array.isArray(body) || Array.isArray(body.data), 'Expected array');
  });

  await test('POST /api/operations creates operation', async () => {
    const { status, body } = await request('/api/operations', {
      method: 'POST',
      headers: auth,
      body: JSON.stringify({
        type: 'activate',
        target: `77${Math.floor(Math.random() * 10000000)}`,
        operator: 'Yemen Mobile',
      }),
    });
    assert(status === 201, `Expected 201, got ${status}`);
    assert(body.type === 'activate', 'Expected type activate');
    assert(body.id, 'Expected id field');
  });

  await test('POST /api/operations with recharge type', async () => {
    const { status, body } = await request('/api/operations', {
      method: 'POST',
      headers: auth,
      body: JSON.stringify({
        type: 'recharge',
        target: `77${Math.floor(Math.random() * 10000000)}`,
        operator: 'Sabafon',
      }),
    });
    assert(status === 201, `Expected 201, got ${status}`);
    assert(body.type === 'recharge', 'Expected type recharge');
  });

  await test('POST /api/operations without target returns 400', async () => {
    const { status } = await request('/api/operations', {
      method: 'POST',
      headers: auth,
      body: JSON.stringify({ type: 'activate' }),
    });
    assert(status === 400, `Expected 400, got ${status}`);
  });

  await test('POST /api/operations without type returns 400', async () => {
    const { status } = await request('/api/operations', {
      method: 'POST',
      headers: auth,
      body: JSON.stringify({ target: '771234567' }),
    });
    assert(status === 400, `Expected 400, got ${status}`);
  });

  await test('POST /api/operations with invalid type returns 400', async () => {
    const { status } = await request('/api/operations', {
      method: 'POST',
      headers: auth,
      body: JSON.stringify({ type: 'invalid', target: '771234567' }),
    });
    assert(status === 400, `Expected 400, got ${status}`);
  });

  await test('POST /api/operations normalizes operator', async () => {
    const { status, body } = await request('/api/operations', {
      method: 'POST',
      headers: auth,
      body: JSON.stringify({
        type: 'activate',
        target: `77${Math.floor(Math.random() * 10000000)}`,
        operator: 'yemen mobile',
      }),
    });
    assert(status === 201, `Expected 201, got ${status}`);
    assert(body.operator === 'Yemen Mobile', `Expected normalized operator, got ${body.operator}`);
  });

  await test('GET /api/inventories returns array', async () => {
    const { status, body } = await request('/api/inventories', { headers: auth });
    assert(status === 200, `Expected 200, got ${status}`);
    assert(Array.isArray(body), 'Expected array');
  });

  await test('PUT /api/inventories updates inventory', async () => {
    const get = await request('/api/inventories', { headers: auth });
    const inventories = get.body.map(inv => ({
      operator: inv.operator,
      available: inv.available || 0,
      remaining: inv.remaining || 0,
    }));
    if (inventories.length > 0) {
      const { status } = await request('/api/inventories', {
        method: 'PUT',
        headers: auth,
        body: JSON.stringify(inventories),
      });
      assert(status === 200, `Expected 200, got ${status}`);
    }
  });

  await test('PUT /api/inventories with empty array returns 400', async () => {
    const { status } = await request('/api/inventories', {
      method: 'PUT',
      headers: auth,
      body: JSON.stringify([]),
    });
    assert(status === 400, `Expected 400, got ${status}`);
  });
}

run();
